DataViews: Support hiding the title in Grid layout, with actions available on hover

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->

Part of #71078 

Update the DataViews Grid layout to support hiding the title, while making the actions available on hover, similar to how selection works.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Discussed in #71078, but overall the goal is to be able to support grid layouts that are simple in appearance (i.e. just a grid of images) while still allowing the flexibility and control of things like actions.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Allow for the `title` field to be hidden
* Only show the `HStack` for the title when the title is present
* If the title isn't present, display the actions over the top of the media preview (if present)
* Fade the selection and actions in on hover so that it feels a little smoother and less heavy
* Update the Storybook examples to ensure that opening the actions menu doesn't cause the padding of the story to jump (this works around a conflict between Ariakit's use of `usePreventBodyScroll` and the default padding applied to the body element in Storybook.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

* Spin up storybook via `npm run storybook:dev` and go to the Default story: http://localhost:50240/?path=/story/dataviews-dataviews--default
* In the options at the top right, hide all the fields except for Image
* Try hovering over each card and interacting with the actions menu
* Also try tabbing through with the keyboard and ensure that it works as expected
* In the config menu for DataViews, click to show the Title field and make everything otherwise works as in `trunk`

## Screenshots or screencast <!-- if applicable -->

<img width="1440" height="987" alt="image" src="https://github.com/user-attachments/assets/6daa4e94-3148-4aa4-a23b-6db27d9f9ab9" />

https://github.com/user-attachments/assets/245f9255-cebe-4376-a786-5b4e78e177e7

andrewserong committed:
DataViews Grid: Try a titleless grid with actions available on hover

andrewserong committed:
Add an explicity is-hovered state to preserve the visibility of the button when popovers are open

andrewserong committed:
Add transition so that it doesn't feel quite as heavy

andrewserong committed:
Update stories so that the padding of the body doesn't jump when opening popovers

andrewserong committed:
Don't force show actions when grid card is selected

andrewserong committed:
Add changelog entry

talldan commented on packages/dataviews/src/dataviews-layouts/grid/style.scss:
```diff
@@ -149,9 +154,41 @@
+	top: $grid-unit-05;
+	opacity: 0;
+	right: $grid-unit-05;
+	z-index: z-index(".dataviews-view-grid__card .dataviews-selection-checkbox");
```
Is the z-index needed? It works ok for me without it.

talldan commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -536,3 +550,43 @@ export const InfiniteScroll = () => {
 	);
 };
+
+export const GridLayoutHiddenTitle = () => {
```
I wonder if we need a separate story, or if it can be an arg on the existing default story 🤔 

There are quite a lot of dataviews stories now.

oandregal replied:
Given the title field would be hidable, any existing story (including the default) would serve to test this flow, without any new story or argument.

I appreciate caring about adding a new story, thanks!

andrewserong replied:
Good points! I'll leave it in for now while I'm testing, but happy to remove it 👍

andrewserong replied:
Removed in https://github.com/WordPress/gutenberg/pull/71369/commits/bd4b9f44a6c872daf6b9192d4bc3f8e721eb6c54

talldan reviewed (commented):
Works mostly well for me in testing. I noted that when using voiceover, the checkbox tells you the title of the item, so the title isn't completely lost, which is good.

When using the story I note that it's now possible to hide absolutely everything and items become completely invisible. I know this was discussed already, but I feel like that's not a great situation, I don't know if there was a conclusion to the discussion:
<img width="1057" height="801" alt="Screenshot 2025-08-28 at 5 24 38 pm" src="https://github.com/user-attachments/assets/975d9f52-57eb-48ca-93c0-e66b0476f390" />

It'd be good to add some tests. In particular a test that checks the checkboxes and actions dropdowns are accessible through the tab key would be great, as it's the kind of thing that ends up as a bug as people tinker with the styles.

talldan commented on packages/dataviews/src/dataviews-layouts/grid/index.tsx:
```diff
@@ -153,6 +154,8 @@ function GridItem< Item >( {
 				infiniteScrollEnabled ? paginationInfo.totalItems : undefined
 			}
 			aria-posinset={ posinset }
+			onMouseEnter={ () => setIsHovered( true ) }
+			onMouseLeave={ () => setIsHovered( false ) }
```
I'm curious if this is needed, as the selection checkbox for items shows/hides itself on hover already and didn't seem to need it.

andrewserong replied:
This one isn't obvious so I should probably add a comment! It's needed, and a similar technique is used elsewhere in Gutenberg (e.g [here](https://github.com/WordPress/gutenberg/blob/eea73800953b713a462384733eb867be7326ab9b/packages/block-editor/src/components/inserter/media-tab/media-preview.js#L258-L263)), because when opening the popover menu, Ariakit's backdrop means that CSS believes we're no longer hovering over the item, and the button becomes invisible. Using state helps us to make the hover behaviour more "sticky" 😄

tellthemachines replied:
Do we control that CSS? Could we attach the hover styles to [aria-expanded="true"]?

andrewserong replied:
> Do we control that CSS? Could we attach the hover styles to [aria-expanded="true"]?

Oh, great tip both of you! I thought it wasn't possible, but I forgot about the `:has()` selector, I think I can get it working without the state now 👍

andrewserong replied:
Yes, that seemed to do the trick (https://github.com/WordPress/gutenberg/pull/71369/commits/524e961bf93ff20bf514ab2222c6012e8b74a66d). Thanks again!

talldan commented on packages/dataviews/src/dataviews-layouts/grid/style.scss:
```diff
@@ -149,9 +154,41 @@
+	}
+
+	@media (hover: none) {
+		// Show checkboxes on devices that do not support hover.
```
Nit - it's not a checkbox.

talldan commented on packages/dataviews/src/dataviews-layouts/grid/style.scss:
```diff
@@ -149,9 +154,41 @@
+	}
+}
+
+.dataviews-view-grid__card:hover .dataviews-view-grid__hover-actions,
+.dataviews-view-grid__card:focus-within .dataviews-view-grid__hover-actions {
+	opacity: 1;
+}
+
+// Include an explicit .is-hovered class to preserve hovered state when popovers are active.
+.dataviews-view-grid__card .dataviews-view-grid__hover-actions.is-hovered {
+	opacity: 1;
+}
```
May also want to consider ensuring `opacity: 1` is used when `aria-expanded` is true on the opener button. At the moment the button will hide itself. `focus-within` doesn't work consistently because the dropdown menu content isn't rendered in the same part of the DOM.
<img width="325" height="333" alt="Screenshot 2025-08-28 at 5 37 25 pm" src="https://github.com/user-attachments/assets/5514cc1f-6730-4abf-835c-6cc3954f894f" />

I also wonder if the selection checkbox should have matching visibility logic, so all the interactive elements for an item are always visible at the same time - e.g. whenever the dropdown is open the checkbox is also visible. Though the table layout's UI doesn't work like this so maybe not. It's a very minor thing so maybe not worth sweating about!

oandregal commented on packages/fields/src/fields/title/index.ts:
```diff
@@ -18,7 +18,7 @@ const titleField: Field< CommonPost > = {
 	getValue: ( { item } ) => getItemTitle( item ),
 	render: TitleView,
-	enableHiding: false,
+	enableHiding: true,
```
I understand we want to display actions in the media field when title is not present. But, do we also want the title field to be hidable in the existing views (site editor's pages)? Not anything against it, just double-checking.

andrewserong replied:
> But, do we also want the title field to be hidable in the existing views

Yes, it'll be a key part of building out [a new media library](https://github.com/WordPress/gutenberg/issues/55238) that's powered by data views, as we'll need to be able to hide the title for attachment post types in a grid view.

oandregal replied:
Yeah, that sounds a good experience for the media library, but can this be addressed differently without affecting existing views? For example, with this view config for the media library:

```js
const view = {
  type: 'grid',
  // titleField not provided
  mediaField: '...',
  fields: [ /**/ ]
};
````

That way, we wouldn't enable hiding the `title` in the site editor's page.

andrewserong replied:
> but can this be addressed differently without affecting existing views?

No, I don't think so. For the media use case, we still want the title field to exist, as it's needed for accessibility (i.e. the label provided to the selection checkbox). So it's important that we can _hide_ the title field, without excluding it altogether from the view.

Also, I think it should still be possible for folks to show the title field if they want to, with it defaulted to hidden.

To mitigate the issue for existing views, that's one of the reasons I was proposing making it so that if any locked fields are present, we can't hide the last one that's visible. So we give users more flexibility, but add a guardrail so that they can't get to an unusable state where no fields at all are visible.

andrewserong replied:
> That way, we wouldn't enable hiding the title in the site editor's page.

Oh, one detail to mention here is that this PR doesn't affect the pages list in the site editor as that uses a different field ([pageTitleField](https://github.com/WordPress/gutenberg/blob/95e8e91a3d2c0d274336ebf8342a98419728c4cd/packages/fields/src/fields/page-title/index.ts#L14)), rather than the generic `titleField` which is used as a fallback here: https://github.com/WordPress/gutenberg/blob/dd70c03c9eae9fb0806e27cd390227a01f428ad3/packages/editor/src/dataviews/store/private-actions.ts#L192

tellthemachines replied:
I think it makes sense for the base title to default to enable hiding. If fields can usually be hidden, _disabling_ hiding is a decision which should probably be made at the consumer level. Both Pages and Templates have their own title fields which disable hiding, so nothing will change in the actual site editor.

oandregal replied:
> Oh, one detail to mention here is that this PR doesn't affect the pages list in the site editor as that uses a different field

Oh, nice. My worry was about affecting existing screens 👍 

Little thing (default value is `true`):

```suggestion
```

> disabling hiding is a decision which should probably be made at the consumer level

I think of `wordpress/fields` as consumer-level code! It's just a different package, but I see it as an opinionated fields bundle for WordPress entities that we expect to use in our screens. While any screen can modify the field, in this bundle, it's best to have good defaults.

Thanks to you both for the context. I now understand this change better.

oandregal replied:
Actually, this field is used in the post list experiment ("Gutenberg > Experiments > enable for posts", then visit "Gutenberg > Posts"). Now that I've looked more into how fields are organized, would this work for you instead?

- Leave the existing `title` field as it is.
- **When** it comes to use the title field in the media library, revisit this. We could/may need to add a `mediaTitleField` that has specific field configs. I understand we don't need that now, but this change was more of a future thing, right?

This would prevent making the field hidable for any CPT entity, for example.

In any case, if you still think it's best to carry on with this change, I'd suggest at least adding a changelog entry in the fields package so consumers are aware of this.

andrewserong replied:
Thanks for the continued discussion here, I do think it's really good for us to think through these nuances!

My take on it is that the base title field should be the most flexible, with each particular instance that inherits from it being more opinionated. Ideally, I don't think we'd create a separate title field for media unless we have other requirements for how the title is displayed than what's included in the base title field. Or alternately, we might have `postTitle` in addition to `mediaTitle`, keeping `titleField` as the generic / base field.

For now, I'm pretty sold on enabling hiding on this field, so I've added a changelog entry for it. As we explore follow-ups to tackle the UX issues with hiding fields, I'm very happy to revisit this, though!

Once this PR lands, let's keep the discussion going on the linked issue: https://github.com/WordPress/gutenberg/issues/71078

> Little thing (default value is true):

In this case, due to the discussion surrounding hiding, I actually prefer keeping the value explicit here so that if someone looks to make a change to it, they'll see that it was added here intentionally.

oandregal commented on packages/dataviews/src/dataviews-layouts/grid/index.tsx:
```diff
@@ -176,24 +179,40 @@ function GridItem< Item >( {
-					renderItemLink={ renderItemLink }
-					className="dataviews-view-grid__title-field dataviews-title-field"
-					{ ...titleA11yProps }
+			{ ! showTitle && showMedia && !! actions?.length && (
```
If the media field doesn't provide a render function, and it's of type `media`, there won't be a `renderedMediaField`. It seems we have to add that check here, the same we do in the sections above.

andrewserong committed:
Update comment to refer to actions instead of checkboxes, ensure opacity is set correctly

tellthemachines reviewed (commented):
This is working well apart from the scenario Dan mentioned:

> When using the story I note that it's now possible to hide absolutely everything and items become completely invisible. I know this was discussed already, but I feel like that's not a great situation, I don't know if there was a conclusion to the discussion:

Could we maybe ensure the title always displays if everything else is hidden?

andrewserong committed:
Try removing state and using aria-expanded instead

andrewserong committed:
Add extra renderedMediaField check

andrewserong committed:
Add tests

andrewserong commented:
Thanks for the great feedback, everyone! I think I've implemented most of the feedback, and I've added a couple of minimal tests to cover tabbing to the checkbox and actions, but let me know if that's not what you had in mind. I found writing the tests a little fiddly, so not sure if they're quite the right way to go about it.

> When using the story I note that it's now possible to hide absolutely everything and items become completely invisible. I know this was discussed already, but I feel like that's not a great situation, I don't know if there was a conclusion to the discussion:

Agreed, and it's being discussed over in https://github.com/WordPress/gutenberg/issues/71078#issuecomment-3227272428. I was keen to try to keep dealing with that separate from this change as I wasn't sure how complex it would be to implement, but I do understand if folks think it's a blocker to landing this one!

I might not get to fixing that up today, but what I'd like to (somehow) try, is adding logic to the Appearance area so that you actually can't hide every property. I.e. disabling the `Hide` button if the field is the last visible field, or something like that. I'll give it a shot, and happy to try that either here, or in another PR.

In the meantime, I'll leave the Story in this PR for now, but will remove it before this PR eventually lands.

Thanks again, all!

andrewserong commented:
Tiny update: after thinking it over and hacking around, I think we _should_ resolve the issue of folks accidentally hiding all fields within this PR. I've got something working locally that looks like this:

https://github.com/user-attachments/assets/fce2c74f-7202-4f09-8237-266f8435fe14

Basically, for each of the "locked" fields (title, preview/media, description) if it's the last visible field, we treat is as locked. It feels pretty good to me so far. I'll likely run out of time to commit the code change as I've still got to handle previews properly, but I think it's promising.

I'll continue on with it next week!

oandregal commented:
> Basically, for each of the "locked" fields (title, preview/media, description) if it's the last visible field, we treat is as locked

Commented in the other thread, resharing for visibility https://github.com/WordPress/gutenberg/issues/71078#issuecomment-3242353511

andrewserong committed:
Remove custom story

andrewserong commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -46,6 +46,20 @@ import './style.css';
 const meta = {
 	title: 'DataViews/DataViews',
 	component: DataViews,
+	// Use fullscreen layout and a wrapper div with padding to resolve conflicts
+	// between Ariakit's Dialog (usePreventBodyScroll) and Storybook's body padding
+	// (sb-main-padding class). This ensures consistent layout in DataViews stories
+	// when clicking actions menus. Without this the padding on the body will jump.
+	parameters: {
+		layout: 'fullscreen',
+	},
+	decorators: [
+		( Story ) => (
+			<div style={ { padding: '1rem' } }>
+				<Story />
+			</div>
+		),
+	],
```
While this PR doesn't add any additional stories anymore, I'm leaving this change in, as it helps make things more consistent when interacting with the actions popovers within any of the DataViews stories.

oandregal replied:
Thanks, that's nice.

andrewserong commented:
> To confirm, this PR only affects the Grid layout, right?

This PR affects the Grid Layout, and the generic `titleField` which is used as a fallback title field for entities (i.e. if we're building a media library using DataViews) [here](https://github.com/WordPress/gutenberg/blob/dd70c03c9eae9fb0806e27cd390227a01f428ad3/packages/editor/src/dataviews/store/private-actions.ts#L192). It will allowing hiding of this field even in other views/layouts.

> I guess I'm really asking, what cannot be done later in a follow up, or what would this PR negatively affect, if it were merged?

I don't believe this PR will have any negative impact within Gutenberg, but consumers that use the generic `titleField` might find that users can now toggle the title field to be hidden (which is one of the desired goals of this PR), even if the view is not Grid. It won't affect the default visibility state, though.

My hunch is that most of the improvements to the UX surrounding when no fields are hidden (or adding guardrails to prevent that from happening) could happily happen in follow-ups.

tellthemachines commented on packages/dataviews/src/test/dataviews.tsx:
```diff
@@ -510,6 +510,92 @@ describe( 'DataViews component', () => {
+				screen.getAllByRole( 'button', { name: 'Actions' } )[ 0 ]
+			).toHaveFocus();
+		} );
+
```
Not something to address in this PR but these feel like they should be e2e tests. Could we perhaps run e2e against storybook for dataviews UI testing?

talldan commented on packages/dataviews/src/dataviews-layouts/grid/index.tsx:
```diff
@@ -176,24 +176,42 @@ function GridItem< Item >( {
+				showMedia &&
+				renderedMediaField &&
+				!! actions?.length && (
+					<div className="dataviews-view-grid__hover-actions">
```
Very minor, but I wonder if the classname should be revised to something like `dataviews-view-grid__media-actions` now that there are two sets of actions.

andrewserong replied:
Thanks for the suggestion, that naming feels clearer (added in https://github.com/WordPress/gutenberg/pull/71369/commits/0a5ca80c43519f352dc55fd1339c5c1258d6ee8f)

talldan commented on packages/dataviews/src/dataviews-layouts/grid/index.tsx:
```diff
@@ -176,24 +176,42 @@ function GridItem< Item >( {
-				{ !! actions?.length && (
-					<ItemActions item={ item } actions={ actions } isCompact />
+			{ ! showTitle &&
+				showMedia &&
+				renderedMediaField &&
```
Small code quality change might be to group all the media related elements behind the same condition (`showMedia && renderedMediaField`). There are a couple of elements above this that also have the same conditional logic.

andrewserong replied:
Good idea, I've tidied these up in https://github.com/WordPress/gutenberg/pull/71369/commits/46ad2a2bcb486ccd81cfef672e895125b553414f

tellthemachines reviewed (approved):
This LGTM as a first iteration! Then we can perhaps follow up with some UI tweaks to discourage doing anything too weird like hiding all the fields at once 😅

talldan reviewed (approved):
This PR in isolation is working well for me and doesn't introduce any issues in the gutenberg plugin. I left a couple of small comments, but they're just code quality things.

I do think it'd be good to have a pretty quick follow-up to prevent being able to hide all fields. I leave it up to you how to handle it, but the solution you showed looks good to me!

andrewserong committed:
Update classname to be a little more specific

andrewserong committed:
Tidy up some boolean checks

andrewserong committed:
Add fields package changelog entry

andrewserong commented:
Thanks for all the reviews and discussion here, folks! I'll merge this in now, and will continue the discussion surrounding the hiding UX over in https://github.com/WordPress/gutenberg/issues/71078 and in follow-ups.
